File: electron/ingest.js
const fs = require('fs-extra');
const path = require('path');
const { probeMedia } = require('./metadata');
const ffmpeg = require('fluent-ffmpeg');
const { configureFfmpeg } = require('./ffmpeg');

// Configure FFmpeg paths
configureFfmpeg();

/**
 * Generate unique asset ID
 */
function generateAssetId() {
  const timestamp = Date.now();
  const random = Math.floor(Math.random() * 1000);
  return `asset_${timestamp}_${random}`;
}

/**
 * Generate thumbnail for video file
 */
async function generateVideoThumbnail(videoPath, outputPath) {
  return new Promise((resolve, reject) => {
    ffmpeg(videoPath)
      .screenshots({
        timestamps: ['00:00:01'],
        filename: path.basename(outputPath),
        folder: path.dirname(outputPath),
        size: '320x180'
      })
      .on('end', () => resolve(outputPath))
      .on('error', (err) => reject(err));
  });
}

/**
 * Generate thumbnail for image file (resize)
 */
async function generateImageThumbnail(imagePath, outputPath) {
  return new Promise((resolve, reject) => {
    ffmpeg(imagePath)
      .outputOptions([
        '-vf', 'scale=320:180:force_original_aspect_ratio=decrease,pad=320:180:(ow-iw)/2:(oh-ih)/2'
      ])
      .output(outputPath)
      .on('end', () => resolve(outputPath))
      .on('error', (err) => reject(err))
      .run();
  });
}

/**
 * Determine asset type from file extension
 */
function getAssetType(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  const videoExts = ['.mp4', '.mov', '.avi', '.mkv', '.webm', '.m4v'];
  const audioExts = ['.mp3', '.wav', '.aac', '.flac', '.ogg', '.m4a'];
  const imageExts = ['.jpg', '.jpeg', '.png', '.gif', '.bmp', '.webp'];
  
  if (videoExts.includes(ext)) return 'video';
  if (audioExts.includes(ext)) return 'audio';
  if (imageExts.includes(ext)) return 'image';
  return 'unknown';
}

/**
 * Ingest files - reference them directly without copying
 * This allows the renderer to access files in their original location
 */
async function ingestFiles(filePaths, cache) {
  const results = [];

  for (const filePath of filePaths) {
    try {
      // Validate file exists
      const exists = await fs.pathExists(filePath);
      if (!exists) {
        throw new Error(`File does not exist: ${filePath}`);
      }

      // Check if it's a file
      const stats = await fs.stat(filePath);
      if (!stats.isFile()) {
        throw new Error(`Path is not a file: ${filePath}`);
      }

      // Extract original file name
      const originalFileName = path.basename(filePath);

      // Generate unique asset ID
      const assetId = generateAssetId();

      console.log(`Ingesting file: ${filePath}`);
      console.log(`Original name: ${originalFileName}`);

      // Ensure cache directory exists (for thumbnails)
      await fs.ensureDir(cache.thumbDir);

      // Get file size from original location
      const fileSize = stats.size;

      // Extract metadata from original file
      const metadata = await probeMedia(filePath);

      // Generate thumbnail (stored in cache, but media stays in original location)
      let thumbnailPath = null;
      const assetType = getAssetType(filePath);
      
      try {
        if (assetType === 'video') {
          const thumbnailFilename = `${assetId}.jpg`;
          thumbnailPath = path.join(cache.thumbDir, thumbnailFilename);
          await generateVideoThumbnail(filePath, thumbnailPath);
          console.log(`Thumbnail generated: ${thumbnailPath}`);
        } else if (assetType === 'image') {
          const thumbnailFilename = `${assetId}.jpg`;
          thumbnailPath = path.join(cache.thumbDir, thumbnailFilename);
          await generateImageThumbnail(filePath, thumbnailPath);
          console.log(`Thumbnail generated: ${thumbnailPath}`);
        }
      } catch (thumbError) {
        console.warn(`Failed to generate thumbnail for ${originalFileName}:`, thumbError);
        // Continue without thumbnail
      }

      results.push({
        asset_id: assetId,
        file_path: filePath, // Use original file path directly
        original_file_name: originalFileName,
        thumbnail_path: thumbnailPath,
        file_size: fileSize,
        metadata,
      });
    } catch (error) {
      console.error(`Error ingesting file ${filePath}:`, error);
      throw error;
    }
  }

  return results;
}

module.exports = {
  ingestFiles,
};

File: src-tauri/src/ingest.rs
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use tauri::{AppHandle, Manager};
use tokio::fs;

use crate::cache::CacheDirs;
use crate::metadata::probe_media;
use crate::ffmpeg::resolve_ffprobe_path;

#[derive(Debug, Serialize, Deserialize)]
pub struct IngestResult {
    pub asset_id: String,
    pub file_path: String,
    pub metadata: crate::MediaMeta,
}

#[derive(Debug, Deserialize)]
pub struct IngestRequest {
    pub file_paths: Vec<String>,
}

/// Ingest files from external paths into the app's cache directory
/// This handles copying files and extracting metadata
#[tauri::command]
pub async fn ingest_files(
    app: AppHandle,
    request: IngestRequest,
) -> Result<Vec<IngestResult>, String> {
    let cache_dirs = CacheDirs::new(&app).map_err(|e| e.to_string())?;
    let ffprobe = resolve_ffprobe_path(&app)?;
    
    // Ensure the media directory exists
    fs::create_dir_all(&cache_dirs.media_dir)
        .await
        .map_err(|e| format!("Failed to create media directory: {}", e))?;
    
    let mut results = Vec::new();
    
    for file_path in request.file_paths {
        let source_path = Path::new(&file_path);
        
        // Validate that the file exists and is readable
        if !source_path.exists() {
            return Err(format!("File does not exist: {}", file_path));
        }
        
        if !source_path.is_file() {
            return Err(format!("Path is not a file: {}", file_path));
        }
        
        // Generate a unique filename to avoid conflicts
        let file_name = source_path
            .file_name()
            .ok_or_else(|| format!("Invalid file name: {}", file_path))?
            .to_string_lossy();
        
        let asset_id = generate_asset_id();
        let extension = source_path
            .extension()
            .and_then(|ext| ext.to_str())
            .unwrap_or("");
        
        let cached_filename = format!("{}.{}", asset_id, extension);
        let cached_path = cache_dirs.media_dir.join(&cached_filename);
        
        // Copy the file to the cache directory
        fs::copy(source_path, &cached_path)
            .await
            .map_err(|e| format!("Failed to copy file {}: {}", file_path, e))?;
        
        // Extract metadata using ffprobe
        let metadata = probe_media(&ffprobe, cached_path.to_string_lossy().as_ref())
            .await
            .map_err(|e| format!("Failed to extract metadata for {}: {}", file_path, e))?;
        
        results.push(IngestResult {
            asset_id,
            file_path: cached_path.to_string_lossy().to_string(),
            metadata,
        });
    }
    
    Ok(results)
}

/// Generate a unique asset ID
fn generate_asset_id() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let timestamp = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis();
    format!("asset_{}", timestamp)
}

/// Get file type from extension
pub fn get_file_type_from_path(path: &Path) -> Result<String, String> {
    let extension = path
        .extension()
        .and_then(|ext| ext.to_str())
        .unwrap_or("")
        .to_lowercase();
    
    match extension.as_str() {
        // Video formats
        "mp4" | "mov" | "avi" | "mkv" | "webm" | "m4v" => Ok("video".to_string()),
        // Audio formats
        "mp3" | "wav" | "aac" | "flac" | "ogg" | "m4a" => Ok("audio".to_string()),
        // Image formats
        "jpg" | "jpeg" | "png" | "gif" | "bmp" | "webp" => Ok("image".to_string()),
        _ => Err(format!("Unsupported file type: {}", extension)),
    }
}
